src/btt.c

```diff
@@ -1060,10 +1181,212 @@ write_layout(struct btt *bttp, int lane, int write)
+		state[i].btt_info_state = BOTH_CORRUPTED;
+
+		/* read primary info */
+		struct btt_info info;
+		state[i].primary_info_off = arena_off;
+		if ((*bttp->ns_cbp->nsread)(bttp->ns, lane, &info,
+					sizeof (info), arena_off) < 0)
+			return -1;
+
+		if (read_info(&info))
+			state[i].btt_info_state = SECOND_CORRUPTED;
+
+		/* read backup info */
+		struct btt_info backup_info;
+		state[i].backup_info_off = arena_off + state[i].size -
+				sizeof (struct btt_info);
+
+		if ((*bttp->ns_cbp->nsread)(bttp->ns, lane, &backup_info,
+			sizeof (backup_info), state[i].backup_info_off) < 0)
+			return -1;
+
+		if (read_info(&backup_info)) {
+			if (state[i].btt_info_state == SECOND_CORRUPTED) {
+				if (!memcmp(&info, &backup_info, sizeof (info)))
+					state[i].btt_info_state = BOTH_OK_IDENT;
+				else
+					state[i].btt_info_state = BOTH_OK_DIFF;
+			} else {
+				state[i].btt_info_state = FIRST_CORRUPTED;
+			}
+		}
+
+		LOG(3, "arena %d state %d", i, state[i].btt_info_state);
+		arena_off += state[i].size;
+	}
+
+	return 0;
+}
+
+/*
+ * copy_info -- (internal) copy src btt_info onto dst
+ *
+ * This function is used only during the initialization of the pool. As such
+ * it does not employ any synchronization mechanism and is not multi-thread
+ * safe.
+ */
+int
+copy_info(struct btt *bttp, int lane, off_t src_off, off_t dst_off)
+{
+	LOG(3, "bttp %p lane %d src_off %lu dst_off %lu", bttp, lane,
+			src_off, dst_off);
+
+	struct btt_info info;
+	if ((*bttp->ns_cbp->nsread)(bttp->ns, lane, &info, sizeof (info),
+			src_off) < 0)
+		return -1;
+	if ((*bttp->ns_cbp->nswrite)(bttp->ns, lane, &info, sizeof (info),
+			dst_off) < 0)
+		return -1;
+	return 0;
+}
+
+/*
+ * fix_arena -- (internal) attempt to repair the arena
+ *
+ * Depending on the state of the arena, there are a few ways to repair an arena.
+ * If an arena is in the SECOND_CORRUPTED or BOTH_OK_DIFF state,
+ * the primary info data is copied onto the backup info.  If an arena is in
+ * the FIRST_CORRUPTED state, the backup info data is copied onto the primary
+ * info. If an arena is in the BOTH_CORRUPTED state, a default info layout with
+ * the BTTINFO_FLAG_ERROR set is written out as both primary and backup info.
+ * Arenas in the BOTH_OK_IDENT state do not require any action.
+ */
+static int
+fix_arena(struct btt *bttp, int lane, const struct arena_info *arena)
+{
+	LOG(3, "bttp %p lane %d", bttp, lane);
+	struct btt_info info;
+	/* necessary for default layout calculation */
+	uint32_t internal_lbasize;
+	int flog_size;
+	uint64_t arena_rawsize = arena->size;
+	uint64_t arena_datasize;
+	uint64_t internal_nlba;
+	uint64_t external_nlba;
+	uint64_t mapsize;
+	uint64_t infooff = arena->size - sizeof (info);
+	uint64_t flogoff;
+	uint64_t mapoff;
+
+	switch (arena->btt_info_state) {
+		case BOTH_OK_IDENT:
+			break;
+		case SECOND_CORRUPTED:
+		case BOTH_OK_DIFF:
+			/* copy primary onto backup */
+			if (copy_info(bttp, lane, arena->primary_info_off,
```

Comment:
In this case you don't check for a < 0, In other cases, specifically calls to write_layout() or read_layout() you do. I can't see a difference that would result in different handling of the return values? If there is a reason let me know. If not, I would prefer you are consistent in how you check for errors. My personal opinion is that in all cases rather than doing if(function..) you specifically check the return value, so if (function.. ) != 0, as an example. Its not wrong that you are doing it this way, its just, in my opinion, easier to understand if you check against a specific value. My point is we should be consistent on how we handle checking for return values.